### DISCUSSION TITLE
FE 로또 미션 구조 토론 | 2022.02.25.

### DISCUSSION THREAD

**Discussion by woowapark** [woowacourse/retrospective#7] [General]
### 로또 미션 1단계를 구현하며 설계해본 과정을 자유롭게 나누어주세요. 
- 우리는 이렇게 설계했다! 를 상대 페어에게 설명해주기
- 어떤 순서로 문제에 접근했는지 이야기해보고, 그 순서 중에 중요한 순간이 언제였는지 이야기해보기

### 체크사항
- ✅ '어떤 생각과 고민의 과정을 거쳤는 지'가 잘 드러날 수 있게 이야기를 나누고 글로도 적어주세요.
- ✅  두 페어(페어 vs 페어)당 1개의 코멘트에 함께 논의한 내용을 작성하여 공유해주세요.

**Comment by prefer2** [comment]
# 안&코이 - 꼬재&도리
## 꼬재&도리
- 미션 요구사항에 집중을 해서 TDD를 중점으로 진행함
- Jest로 단위 테스트 -> Cypress도 이전에 공부했으니 다시 사용해서 E2E테스트도 진행해봄
- 우선은 하나의 Class에 test단위로 구현 -> 이후 UI, 함수 등을 분리함(하다보니 MVC 느낌)
- 구조에 대해서도 고민했었지만 시간적 한계로 따로 구조를 진행하지는 못함

## 안&코이
- 자연스럽게 MVC....
- 패턴에 따라 분리가 잘 되어서 좋았음
- 하다보니 MVC에 중점을 두고 미션을 진행하게 되었음(아쉽)
- 단위 테스트를 어떻게 해야할지 고민을 하게 됨 -> 코드를 먼저 진행하고 테스트코드를 작성하여 어려움이 있었다

## 결론
- MVC에서 View를 왜 클래스로 관리할 필요가 있을까? View가 한개라면 함수로 진행해도 괜찮지 않은가?-> View를 섹션별로 나눈다면 Class화 시켜도 좋겠다. 아니라면 굳이 Class화를 해야할까? 
- 질문 잘하는 법, 어떻게 어떻게 진행해서 이렇게 저렇게 했는데 어떤가요? 아애 미션을 진행하면서 있었던 어려움을 그때그때 적어놓고 질문을 고민하는 것도 좋을 것 같다.

**Comment by onschan** [comment]
# 온스타 & 동키콩 vs 코카콜라 & 하리 & 록바

## MVC 를 쓴 이유와 안 쓴 이유

### MVC를 쓴 이유

- 페어, 일정의 기간이 있다. 메뉴얼을 정해야한다. 코드를 안정적으로 작성해야하는데 셋이서 다 같이 이해하고 있는 패턴이 MVC 패턴이었다.

- 3인페어, 각자가 추구하는 것을 하면 합의점을 찾는 것이 어렵다. 친숙한 것이 MVC. MVC 가 좋다고 생각한 이유가 MVC로 짜보고 MVC가 아닌 것으로도 짜봤다. 남들이 봤을 때, 리뷰어가 봤을 때 이해하기 쉬운 코드가 MVC 라고 생각했다.
리뷰어가 읽기 쉬운 코드를 지향해서, MVC 코드를 작성했다.

- 익숙한 패턴을 쓰는게 의견 조율에 있어서 더 낫고, 시간이 없었기 때문에 패턴을 정했다.
공통적으로 “패턴을 정해야한다” 라는 강박이 있었던 것 같다.
하지만 돌아가도 MVC 패턴을 쓸 것같다.

### MVC를 안 쓴 이유

- 처음에 MVC를 안 쓴 이유는 프리코스때부터 MVC를 계속 사용하다보니깐 이번 미션에는 기초로 돌아가서 UI와 도메인만을 분리하자라는 생각을 했기 때문에 패턴을 배제하고 코딩했다. 리팩토링이라는 책을 읽고있는데 구조가 깔끔한 코드를 짜는 것이 아니라 개발자의 능력을 키우는게 우테코의 목적이라 생각하는데, 문제 해결능력이 개발자의 능력인데 문제를 마주치지 않기 위해 해결책을 차용해서 코딩하는 건 그것과 멀어진다 생각한다. 먼저 문제를 부딪혀야 한다고 생각한다.

- MVC에 대한 이해가 바탕이 되어있지 않은채, 남들을 따라하는 것같아 지양하자고 생각함. 준과 포비가 했던 방식처럼 jest를 이용한 TDD로 시작을 해서 UI 없이도 동작하는 도메인만을 만들고 이후에 UI를 구현해보자고 페어와 합의를 하고 페어 프로그래밍을 진행했다. 코드를 전체적으로 더 넓은 시각으로 바라보고 생각할 수 있었다고 생각한다.

---

## MVC를 썻을 때와 안썻을때 차이가 있었나?

- MVC를 사용하면 편하긴 하다. 구조적인걸 생각안하고 따라가기만 하면 되니깐.
뷰의 역할, 모델의 역할, 컨트롤러의 역할 정해져있기때문에
분리를 해보면서 코드적인 (클래스 분리 같은) 부분에 대해서 더 고민할 수 있는 기회가 있었던 것 같다.

- 크게 효용성은 못느꼇다. 하지만 구조적인 부분보다 우리가 중점적으로 생각하는 것에 더 집중할 수 있었던 것같다.

---

## 다음 미션땐 어떻게 할것같냐?

- 컴포넌트를 쓸것같다.

- 최대한 다음 미션에는 MVC를 버릴 것같다.

- 나는 도메인을 만들고 UI를 만드는 과정을 다시 거쳐볼 것같다.

- 패턴을 버려야 하지 않나 생각을 하고있다. 다음번 페어랑 같이 얘기를 해봐야할 것같다. 

- 문제해결 능력이 굉장히 중요하다고 생각, 문제 인식이 먼저라고 생각한다. 제가 하고싶은 건 맞코딩을 해보고 그 다음에 구조를 생각해보는게 좋지 않을까 생각한다. 리팩토링을 빠르게 짧은 주기로 하면서 구조를 맞춰갈 것같다.

---

## 상속을 왜 안썼냐?

- 필요성을 못 느끼고 익숙하지 않아서 안썼던 것같다.

- 상속을 잘 못하면 의존성이 생겨서 오히려 불편하고 위험할 수 있다고 들었다. 알아보고 싶다.

- 몰라서 안썼던게 컸던거같다.

- 화면 구성이 단조로워서 안썼다.

---

## 바닐라로 짜는 컴포넌트
- 코카콜라의 프리코스 자판기 미션 참조
[코카콜라 프리코스 자판기 미션](https://github.com/intae92/javascript-vendingmachine-precourse/tree/intae92)

---

## 록바의 고찰

- 미션 제출하고 코드 제 스스로 코드를 보면서 고치고싶은 부분이 많았다.
어떻게 고칠지 생각했다. 
코치분들과 리뷰어분께 패턴에 집착하지말라는 얘기를 듣고 고민을 많이 했다.
‘개발자란 무엇일까?’ 
Object.freeze 같은 경우도 남들이 쓰고 있는 거에 대해서 좀 따라하는 경향이 있었다.
남들이 하는 걸 나도 하는 경향이 많았다.
이제는 나만의 길을 걷겠다. 

---

## 왜 함수형 프로그래밍이 아닌 클래스로만 이루어진 프로그래밍을 한다고 생각하는가?

- ES6에 클래스가 생겼고 리액트도 그렇게 될 수 있을거같다고 생각

- 우테코 프리코스 PDF 에서 클래스 쓰라고해서 썼다.

- 좀 더 많은 고찰이 필요할 것같다.

**Comment by greenblues1190** [comment]
# 위니&자스민, 밧드&우디

얘기를 나누고 나서 두 페어 모두 왜 써야하는지 정확히 알지 못하고 타성에 젖어 패턴을 사용한 것 같아요. 하지만 뷰와 도메인 로직을 나눠야한다는 건 의견이 같았습니다. 또 MVC의 각 모듈의 역할을 어떻게 나눌 것인가에 대해서도 서로 조금씩 달랐습니다.

다음은 서로 달랐던 부분입니다.

### 위니&자스민

- MVC로 나눈 것 뿐만 아니라 뷰도 역할에 따라 분리했습니다.
- model에서 유효성 검사 후 컨트롤러에서 에러 핸들링을 했습니다.

### 밧드&우디

- 뷰를 하나로 구성하고 템플릿을 컴포넌트 단위로 분리했습니다.
- 뷰에서 유효성 검사를 하고 검증된 입력값을 정제해 컨트롤러에게 넘깁니다.
- 모델 뿐만 아니라 뷰에서만 필요한 상태는 뷰에서 관리하도록 했습니다.

---

패턴이 정답이 아니라는 점에 공감하고 포비가 말했던 직접 패턴을 발견하는 순간을 경험해보고 싶어졌어요. 자스민은 오늘 바로 패턴을 찾기 위한 여정을 시작할 예정이라고 합니다! 우디는 앞으로 MVC를 안 쓸 거라고 하네요. 과연 지킬 수 있을지...

**Comment by jhy979** [comment]
# 나인 - 비녀 - 시지프 - 호프 페어 토론

## 나인 - 비녀 👍 
 
### 고민
- 모델부터 개발하여 도메인을 먼저 분리했고 UI도 따로 개발하여 합친다는 느낌으로 개발을 진행했습니다. 
- Toggle Switch와 Lotto 텍스트가 겹칠 수 있을 것 같아 Grid를 적용해보자고 생각했습니다.

### 해결
- TDD로 모델을 만들어 도메인과 UI를 분리할 수 있었습니다.
- `grid-area`라는 속성을 이용하여 원하는 위치에 요소를 배치할 수 있었습니다.

### 궁금증
- 다른 페어의 코드를 보면서 가장 먼저 했던게 내 페어와 짠 코드와 비교해보는 것이었습니다.
- 같은 패턴이라도 조금씩 다른 구조에 대해서 생각해 보고 직접 답을 얻을 수 있었습니다.

## 시지프 - 호프  👍 

### 구현 시 고민했던 부분

- 도메인과 UI를 분리하는 것에 중점을 두고 고민을 했었다. 이를 MVC 패턴으로 해결하려고 했다. 
- model 과 view 의 책임을 분리하는 것에 신경을 썼다. 
- 그리고 controller가 model과 view의 내용을 어디까지 알아야할지 많이 고민했다. `추상화`하려고 신경 썼다.
- 구현사항 없이 테스트 코드 짜는 것이 막연했다. 그래서 전체적으로 어떤 함수와 변수를 사용할지 고민해보고 시작하여 해결했다.

### 토론

- view 파일 이름과 class `파일 이름이 달라서 혼동`이 일어났다. 이를 통일하면 좋겠다.
- view 에서 DOM요소를 configureDOM으로 한번에 가져왔는데, DOM을 하나의 메소드로 `한번에 불러올지`, DOM이 `필요할 때마다` 그 메소드에서 선언할지 고민해볼 수 있을 것 같다.
- 다른 페어의 코드를 보고 이야기하는 것이 의미있었다.

**Comment by yunjin-kim** [comment]
## 결 & 앨버
- 문제를 해결하려기 보단 해결책을 도입했습니다
- Jest로 단위 테스트를 진행했습니다
- view에서 view(UI) 이벤트를 처리하고 controller에서는 최소한으로 view에 대한 코드를 작성했습니다

## 태태 & 블링
- 작년 기수꺼를 살펴보다가 가장 직관적으로 MVC가 구현하기 쉬워보여서 선택했습니다
- 로또를 구매하려고 했을 때 입력 금액을 검증하고 여러 개의 로또를 생성해서 관리하는 역할을 하는 매니저의 필요성이 느껴져서 구현했습니다

## 결론
- MVC 패턴을 타파하는 방법
- 역할이 모호해 보이는 controller부터 축소해보는 방향

**Reply by jhy979** [reply]
MVC 패턴을 타파하는 방법 zzz

**Comment by dayelop** [comment]
# 돔하디 & 민초, 병민 & 티거

## 🔥 돔하디 & 민초 (MVC 패턴 적용)

- UI와 도메인을 분리하고 서로 간섭을 최소화하기 위해서 MVC 패턴을 사용했다. 이렇게 함으로써 도메인과 UI가 파일별로 직관적으로 나눠져서 유지보수가 편리했던 이점이 있었다.
- 단위 테스트 코드를 작성할 때 UI에 의존하지 않는 도메인 고유의 분리된 코드를 작성할 수 있었다.
- 웹팩을 활용해서 HMR을 잘 이용해보고 싶었는데 알 수 없는 이유로 저장과 동시에 빌드가 바로 반영이 안 되는 상태이다. 이 부분은 스텝2를 하면서 계속 해결해나가야 한다.
- step1에서는 기본적으로 제공하는 웹팩 설정을 그대로 사용했는데, 웹팩은 개발 환경에 편의성을 제공하는 도구인 만큼, step2에서는 추가로 커스텀해서 사용해보아야겠다는 생각도 들었다.

## 🔥 병민 & 티거 (Flux 패턴 적용)

- 상태 변화에 따라서 뷰를 그려주고 싶었다.
- 데이터 흐름을 단방향으로 하고 싶었다.
- 상태 관리를 한곳에서 하고 싶었다.
- Flux 패턴을 적용했다.

## 🔥 포코 추천 사이트 (~~포코 이거 아니면 답변 달아주세요🥺~~ 맞다고 하셨습니다)
- https://overreacted.io/
- https://kentcdodds.com/blog

**Comment by usageness** [comment]
# 로또 미션 구조 토론

## 해리&샐리

### 설계 프로세스

요구사항을 문서로 모두 작성 → 로또 구입 금액 입력값과 로또 번호들의 유효하지 않은 케이스 정리 → 앞서 정리한 케이스 대로 테스트 코드 작성 (빈값 확인하는 함수, 중복 값 확인하는 함수 등을 미리 구현) → app.js에서 클릭이벤트를 받는 함수들 부터 구현 시작 → 로또 구입 금액 입력값 처리 함수 만듬 (앞서 테스트 코드에서 구현되었던 함수를 활용) → 구입 금액에 따라서  로또 객체 생성 → 로또 객체에  랜덤한 값 구현 할당 (이때 중복 값 확인하는  함수를 앞서 만들었던 함수로 사용) → 화면에 로또 리스트와 결과 확인하기 버튼 렌더링 → 토글 이벤트 함수 app.js에서 받음 → 해당 toggle에 따라 is-active를 부여하여 is-active일 경우 화면상에 로또 이미지와 숫자들이 보이도록 함 

### 설계 아이디어

저희는 app class에서 이벤트를 handle해주는 메서드를 만들고, core 폴더에서 주로 로직을 처리해주는 함수들을 작성한 뒤, view 폴더에서 render를 해주는 함수들을 작성했습니다. 그리고 app에서 이벤트 단위로 일을 처리해주는 방식으로 구조를 짰는데요

대충 돌아가는 느낌만 코드로 보자면

```jsx
class App {

	handleEvent() {
		//addEventListener로 이벤트 처리하는 함수 호출
	}

	handleXXXX() {
		//addEventListner에게 호출되어서 하나의 이벤트를 담당
		-> 로직 처리해주는 함수 호출 (core폴더)
		-> 렌더해주는 함수 호출 (view폴더)
	}
}
```

### 상대 페어에게 궁금했던 점

- selector를 상수로 분리하여 처리한 점
- lottoList를 deepcopy해서 넘겨주는 이유가 궁금했다.

---

## 준찌&유세지

### 우리의 설계 프로세스

먼저 각자의 아이디어를 나누고 합의된 설계를 문서화 시켰습니다.

이를 바탕으로 도메인 모델이 어떤 형태를 갖게 될지 대략적인 개요를 머릿속에 가진 채로 테스트 코드를 작성하였고, 이를 만족하도록 코드를 작성하는 것으로 `Lotto` 객체를 완성하였습니다.

그 뒤 피그마 시안에 따라 화면을 그리고, 이를 렌더링하는 `LottoGameView` 를 작성하여 데이터를 저장하고 가공하는 역할과 화면에 이를 보여주는 역할을 분리하였습니다.

마지막으로 이 둘을 연결해주는 역할을 `app.js` 에게 부여하여 각각 데이터, 상호작용, 화면으로 세 개의 관심사를 나누었습니다.

### 설계 아이디어

저희는 MVC 패턴을 적용하고자 하지 않았습니다. 설계의 키워드는 `관심사의 분리`입니다. 

도메인과 UI의 분리라는 타이틀이 있었던 만큼, MVC를 적용하여 구현한다 보단 `도메인과 UI를 명확히 분리`해보자를 목표를 두고 설계와 개발에 임하였습니다. **(근데 MVC로 가는건 왜일까요. 파격적인 사고의 전환을 해봐야 할 것 같아요)**

- 피그마 디자인과 기능 요구사항을 보았을 때, 데이터는 두 가지가 있다고 판단 → `LottoModel` - `LottoGameModel`
- `LottoGameManager` → 이름만 컨트롤러가 아니였지  사실 컨트롤러같네요. 모델을 업데이트하고, 뷰에게 업데이트 된 모델을 전달하여 화면 렌더링하는 역할을 합니다.
- `LottoGameView` → 데이터만 받으면 뷰를 그릴 수 있도록 구현하였습니다.

### 상대 페어에게 궁금했던 점

- 이벤트 바인딩 부분에서 인라인으로 이벤트를 작성한 점
    - disbind 이슈로 인해 이벤트를 화살표 함수로 인라인으로 작성함. 다른 방법은 없을까 고민

**Comment by kwannee** [comment]
# 페어 토론

## 후이 & 콤피

- 앱이 확장될 경우를 우선적으로 염두에 두며 구현했다.
    - 이벤트 리스너를 관리하는 것은 뷰에서 할 일이라 생각하여, 뷰에서 이벤트 등록을 처리하고, 컨트롤러의 핸들러를 받아 실행하는식으로 구현하였다.
        
        > 추후 앱에서 다양한 기능을 추가하였을 때를 고려하여
        뷰에서 사용하는 엘리먼트와 전달해주는 값을 명확하게 보여주고 싶었고 뷰와 컨트롤러에서 엘리먼트를 중복하여 선언하는 것은 DOM을 중복적으로 탐색하는 코드를 야기 시킬 수 있었기에 한번 분리해보았다.
        > 
        
    - dataset으로 로또 목록의 스타일이 변경되도록 구현해보았다.
        
        > 유지/보수를 통해 이후 다양한 기능과 스타일이 추가/변경되었을 때
        classList를 통해 class가 많이 추가되었을 때 혼동이 올 수 있을 것이라 생각하여 엘리먼트가 가진 상태를 명확하게 나누고 싶었다.
        > 

- 우연 소피아 두 분의 설계를 보며 우리 코드와는 다른 장점을 느꼈다. 일반적 인간 사고에 자연스럽게 부합하는 로직이었고, 그만큼 이해하기도 편했다.
- 질문을 받았을 때 막히는 부분들이 있었다. 근거 없이 코딩을 했기 때문인 것 같다. 앞으로 설계에서 좀 더 깊은 생각을 해야겠다고 느꼈다.

### 우연 & 소피아

- View 분리를 중점적으로 생각하며 구현하였다.
    - 처음 시작하면서 MVC가 뭔지도 모르겠고, 왜 쓰는지 잘 모르겠다고 이야기했다. 그래서 미션 목표인 UI와 도메인 분리에 집중하였다. UI 작업은 뷰 업데이트 위주라고 생각하여, View에서는 업데이트 동작만 하도록 작성했다.
- event의 경우 view말고 다른 class에서 진행하였다.
    - 후이와 콤피는 view에서 같이 진행하였다.
    - 전체 동작은 한번에 관리할 수 있도록, (lotto 데이터를 관리하는) LottoManager와 View를 LottoMachine에서 가지고 있고, 데이터를 가져오거나 이벤트를 세팅하는 기능은 이 LottoMachine에서 가지도록 하도록 하였다.
    - 후이와 콤피의 코드 설명을 들으면서, View에서 이벤트를 세팅하는 방법을 새롭게 확인할 수 있었다. 각 기능을 컴포넌트(?)로 나누고 각자 수행하는 기능을 컴포넌트에서 이벤트 세팅을 할 수 있겠다는 아이디어를 얻을 수 있었다.
- 결과 화면 구성은 두 화면을 모두 구현한 뒤 display-none이라는 class를 부여하여 토글로 화면을 바꿔주었다.
    - 후이와 콤피는 data-set을 이용하여 화면을 바꿔 주었다.
    - 받았던 피드백 : 나중에 큰 금액이 들어왔을 때 너무 많은 요소들로 인하여 렌더링이  느려지는 문제가 생길 수 있다.
    - 너무 많은 요소들이 추가되어 렌더링이 힘들어 질 수 있다는 점을 생각하지 못했다. 나중에 pagination이나 하나의 화면을 다르게 보여질 수 있도록 구현하면 좋을 것 같다.
    
- 각자 코드를 설명하면서, 내가 왜 이렇게 작성했는지도 돌아보고 새로운 아이디어도 얻을 수 있었다.

**Comment by wonsss** [comment]
# 빅터-아놀드, 무비-마르코
## 빅터-아놀드
- 최대한 간단히 코드를 짜려고 했다. 틀에 얽매이지 않으려고 했는데, 무의식적으로 MVC 패턴과 비슷한 설계가 나와서 틀을 깨는 연습을 계속해야겠다고 느꼈다. 
- MVC 패턴을 안 쓴 이유는 포코가 패턴에 너무 얽매이지 말라고 말씀해주셔서, 이번에는 안 쓰려고 노력했다. 하지만 무비와 마르코의 MVC 패턴 코드를 보다 보니까, MVC 패턴에 장점이 있는 것 같다( ❤️ ) . 
- 클래스를 습관처럼 써왔는데, 이번에는 굳이 인스턴스를 반복해서 생성하는 클래스를 써야만 하는지 고민이 들었다. 그래서 클래스를 대체할 수 있는 간단한 객체나 함수만을 사용하여 코드를 짜보았다. (쉽지 않은 방향성일 수 있었는데 함께 잘 같이 해준 페어 사랑해, 나두 ❤️  ) 

## 무비-마르코
- 처음에 MVC 패턴에 얽매이지 않고 기능에 따라 설계해보자라는 마음을 가졌었는데, 도식화 및 파일 네이밍을 하다보니까 익숙하던 MVC와 다시 만나게 되었던 것 같다. 
- 그래서 일단 쓰게 된 MVC이니까, 잘 써보자라는 생각에, 기존의 우리가 작성했던 MVC 코드보다 더 개선할 점을 찾으려고 했었다. 그러다가 찾은 문제점이...(아래에서 계속)
- 이전 레이싱카 미션에서는 MVC를 쓰긴 했는데, View의 역할들을 Controller가 많이 하는 것 같았다. 그래서 이번에는 View의 역할이 단순히 렌더링하는 것에 그치지 않도록 하고, 이벤트의 발생과 UI에서 인풋 값을 컨트롤러에게 알려주는 역할을 추가했다. 컨트롤러는 이벤트를 직접 바인딩하고 해당 이벤트에 따라 메서드를 호출하는 것이 아니게 만들었다. 이를 위해 커스텀 이벤트 기반으로 이벤트를 핸들링해보았다. 
- 아놀드와 빅터는 관습처럼 MVC패턴 및 심지어 클래스도 사용하지 않고, 객체나 함수만을 갖고 구현한 것을 보고 놀랐다. 당연하게 클래스를 써왔는데, 클래스는 왜 쓰는 것이고 그냥 객체를 사용하는 것보다 어떤 점이 더 나은지 깊게 고민할 수 있게 됐다. 앞으로도 패턴 뿐만 아니라 문법, 패키지 등 다양한 방법들을 맹목적으로 사용하는 것이 아니라 필요성을 찾고 여러 다양한 방법을 직접 구상해봐야겠다. 
- MVC 패턴을 목표로 짰었는데, 다른 분들 설계 및 코치들의 말씀을 듣고 MVC에 맹목적으로 얽매였던 과거의 나에게 반성하게 됐다. 왜 MVC 패턴을 사용하는지 고민부터 해야 했다. 그리고 MVC를 쓰더라도 그 역할이 고정되어 있는 이유가 무엇인지도 고민을 해봐야 했다. 앞으로는 귀납적으로 공부해야 겠다. 굳이 베스트 프랙티스인 결과물을 학습하려고 노력하지 않아야 겠다. 근데 맨땅에서 헤딩..너무 어려울 것 같긴 하다.ㅠ 무의식적으로 또 MVC처럼 짜게 될 것 같다는 생각이... 
- 무비와 함께 한 MVC 여행... 여름이었다..
- 그러나 MVC 여행.. 당분간은 다시 안 갈 예정이다.  끝.
